### PR TITLE
Add list table

### PR DESCRIPTION
**Pull request by wpwood**
# List Table

Tables are described at http://rackerlabs.github.io/canon/ui-components/#tables. A working table example can be found in DemoListTableSection.jsx.

screen shot 2016-07-15 at 11 34 52 am

screen shot 2016-07-15 at 11 35 37 am

List tables and embedded list tables are similar, except for the component wrapping them. The regular list table is not wrapped in a border, while the embedded list table has a fixed size and is wrapped in a border.

## Wrapper components

The main component for a list table is `ListTable`, and the main component for the embedded list table is `EmbeddedListTable`. Both of these components take the following properties:

- `overlayStatus` - takes a value of 'none', 'empty', 'error', or 'loading', and describes what overlay should be shown on the table.
- `emptyOverlay` - an `EmptyOverlay` component describing the table's empty state.
- `errorOverlay` - an `ErrorOverlay` component describing the table's error state.
- `loadingOverlay` - a `LoadingOverlay` component describing the table's loading state.

The `EmbeddedListTable` takes an additional property, `size`, which can take the values 'small', 'medium', and 'large', and which specifies how tall the embedded list table will be. These correspond to the small, medium, and large sizes described in http://rackerlabs.github.io/canon/ui-components/#tables for the embedded list table.

In addition, there is an `OverlayStatus` enum that contains the values `NONE`, `EMPTY`, `ERROR`, and `LOADING`, and which can be used for the `overlayStatus` values.  

## ListTableHeader

The `ListTableHeader` component specifies how the table header is rendered, as well as how sorting is handled. The children of the `ListTableHeader` are assumed to be `th` components or a component that will resolve to a `th` component that describes a specific header cell. For instance, a valid `ListTableHeader` is:

```
<ListTableHeader>
  <th className="rs-table-status" />
  <th className="rs-table-sort-text">Name</th>
</ListTableHeader>
``` 

There are several predefined header cell components that implement behavior that conforms to the `th` behavior described for canon-react. These are:

- `CheckboxColumnHeader`
- `StatusColumnHeader`
- `TextColumnHeader` - implements sortable behavior, calls the onSort callback when clicked, displays the sort direction indicator.

When a collection of objects is passed to the list table, it is assumed to be in the order that the user of the component wants it in. The `ListTableHeader` takes the following properties for handling sorting:

- `onSort` - a callback function that is called with the current sort direction and columnId for a sortable header. Any header that has the `sortable` boolean property will be sortable, although the `TextColumnHeader` is the only predefined header that supports that property out of the box. If the header is sortable, it must also support the `columnId` property.
- `sortColumn` - the `columnId` for the column that the collection is sorted by.
- `direction` - current sort direction, either 'asc' or 'desc'.  There is a `SortDirection` enum with values `ASCENDING` and `DESCENDING`, which may be used for the direction value.

## ListTableBody

The `ListTableBody` component takes a collection of data, and then for each object in the collection, it clones its children and passes the object as the child's `instance` property.

```
const collection = [
  {id: 1, name: 'one', count: 1},
  {id: 2, name: 'two', count: 42},
  {id: 3, name: 'three', count: 12},
  {id: 4, name: 'four', count: 6}
];

<ListTableBody collection={ collection }>
  <ListTableRow data-model_id={ instance.id } instance={ instance }>
    <td className="rs-table-status rs-table-status-ok"></td>
    <td className="rs-table-checkbox"><input type="checkbox" /></td>
    <td className="rs-table-text">{ instance.name }</td>
    <td className="rs-table-text">{ instance.count }</td>
  </ListTableRow>
</ListTableBody>
```

The `ListTableRow` component is a convenience for passing the `instance` property to each of its children, but in this case, one could just as easily have used a `tr`. The `ListTableBody` will also set the `key` property for the cloned children. By default, this will be the `id` of the instance. If the instance does not have a unique `id` property, then another unique property can be supplied by setting the `keyGenerator` property on the `ListTableBody`.

# Overlays

There are three predefined Overlay components, the `EmptyOverlay`, the `ErrorOverlay`, and the `LoadingOverlay`. Each of them is intended as a default overlay for the `ListTable` and `EmbeddedListTable` wrapper components. There is also a base `ListTableOverlay` which each of these predefined overlays uses, and which may be used if the table needs some non-default overlay behavior.

screen shot 2016-07-15 at 11 35 06 am
screen shot 2016-07-15 at 11 35 19 am
screen shot 2016-07-15 at 11 35 28 am

### CODE REVIEW

**Comment by Jigarsolanki** [comment]
Could you add some description and usage of it?

File: src/DataTable.jsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { Component, PropTypes } from 'react';
+
+class DataTable extends Component {
```
Comment (Jigarsolanki) [inline src/DataTable.jsx]: Why do we call it `DataTable`? why not just `Table`?

File: src/DataTable.jsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    } = this.props;
+
+    return (
+      <div { ...this.props } >
+        <table className="rs-list-table">
+          { children }
+        </table>
+        { isEmpty ? emptyOverlay : null }
+        { isLoading ? loadingOverlay : null }
+        { isErrored ? errorOverlay : null }
```
Comment (MACSkeptic) [inline src/DataTable.jsx]: if all three are true (empty, loading and errored), do we render three overlays on top of each other?

should the status prop be a enum instead?

File: src/DataTable.jsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+DataTable.defaultProps = {
+  isEmpty: false,
+  isErrored: false,
+  isLoading: false
+};
+
+DataTable.propTypes = {
+  children: PropTypes.node,
+  className: React.PropTypes.string,
+  emptyOverlay: PropTypes.node,
```
Comment (MACSkeptic) [inline src/DataTable.jsx]: should we make these child components with a specific type instead of being props?

File: src/DataTableBody.jsx [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  render() {
+    return (
+      <tbody>
+        { this._cloneRowPerInstance() }
+      </tbody>
+    );
+  }
+
+  _cloneRowPerInstance() {
+    return this.props.collection.map(this._cloneRow.bind(this));
```
Comment (jdrush89) [inline src/DataTableBody.jsx]: bind in the constructor

File: demo/DemoPopoverSection.jsx [hunk 5]
```diff
@@ -104,7 +104,7 @@ class DemoPopoverSection extends React.Component {
                   <span>
                     To top align the content on a submittable popover declare the 'horizontal' prop as false:
                   </span>
-                  <br>
+                  <br />
```
Comment (hannahstwinter) [inline demo/DemoPopoverSection.jsx]: 😅 Thanks!

File: Gruntfile.js [hunk 6]
```diff
@@ -67,6 +67,7 @@ module.exports = function (grunt) {
           'test_bundle.js': ['test-built/**/*.js']
         },
         options: {
+          exclude: ['react/lib/ReactContext', 'react/lib/ExecutionEnvironment'],
```
Comment (MACSkeptic) [inline Gruntfile.js]: for my benefit: why are those needed?

File: src/DataTableBody.jsx [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  _cloneRow(instance) {
+    return React.cloneElement(this.props.children, {
+      instance: instance,
+      key: instance.id
+    });
+  }
+}
+
+DataTableBody.propTypes = {
+  children: PropTypes.element.isRequired,
```
Comment (hannahstwinter) [inline src/DataTableBody.jsx]: I wonder if this should be a TableRow component instead, implemented with loads of flexibility in terms of structuring. That way it offers a bit of documentation as to intended usage.

File: Gruntfile.js (hunk 6 above)
Comment (wpwood) [inline Gruntfile.js]: They cause build errors when updating enzyme if they're not there. Actually, it looks like I ned to change this, according to https://github.com/airbnb/enzyme/blob/master/docs/guides/webpack.md

File: src/DataTable.jsx (hunk 2 above)
Comment (hannahstwinter) [inline src/DataTable.jsx]: I agree this could use more consideration... A status enum might not be a bad thing, that way you can control what values are passed in and possibly even offer default overlay props so it both documents itself and still works without having to touch all of the parts. Also... will React choke on rendering undefined values? If someone passes in isEmpty as true but no emptyOverlay, what will happen? I like trying to code these with the assumption that anyone using it is going to try to break it. 😜

File: src/DataTableBody.jsx [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+class DataTableBody extends Component {
+  render() {
+    return (
+      <tbody>
+        { this._cloneRowPerInstance() }
+      </tbody>
+    );
+  }
+
+  _cloneRowPerInstance() {
```
Comment (MACSkeptic) [inline src/DataTableBody.jsx]: 👍  clever

I wonder if there are scenarios where this wouldn't be desirable (like certain rows rendering differently enough that you'd want different components)

I guess that could be handled within the row itself

File: src/EmbeddedDataTable.jsx [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    const sizeClass = SIZE_CLASSES[size];
+
+    classes = classNames(
+      'rs-embedded-list-table-wrapper',
+      sizeClass,
+      className
+    );
+
+    return (
+      <div { ...this.props } className={ classes }>
```
Comment (hannahstwinter) [inline src/EmbeddedDataTable.jsx]: Why do you need ```{ ...this.props }``` here?

File: src/DataTableBody.jsx [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    return this.props.collection.map(this._cloneRow.bind(this));
+  }
+
+  _cloneRow(instance) {
+    return React.cloneElement(this.props.children, {
+      instance: instance,
+      key: instance.id
```
Comment (MACSkeptic) [inline src/DataTableBody.jsx]: is "id" always a safe key?

(asking that after I've created quite a few lists with `awsAccountNumber` or something else as the key)

of course we could just map the input to match, or should we accept a function to generate the key?

File: Gruntfile.js (hunk 6 above)
Comment (wpwood) [inline Gruntfile.js]: Oh, we're not using webpack, so I'll mark them as externals in browserify as shown in http://benclinkinbeard.com/posts/external-bundles-for-faster-browserify-builds/

The actual build error is `Error: Cannot find module 'react/lib/ReactContext' from '/Users/bill.woodward/Projects/canon-react/node_modules/enzyme/build'`

File: src/DataTableHeader.jsx [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+      onSort(direction, child.props.columnId);
+    };
+  }
+}
+
+DataTableHeader.defaultProps = {
+  onSort: () => {}
+};
+
+DataTableHeader.propTypes = {
```
Comment (MACSkeptic) [inline src/DataTableHeader.jsx]: are all tables sortable? maybe I missed where one could opt out

File: src/DataTableBody.jsx (hunk 7 above)
Comment (MACSkeptic) [inline src/DataTableBody.jsx]: 👍  big fan of making it a child component instead

File: Gruntfile.js (hunk 6 above)
Comment (MACSkeptic) [inline Gruntfile.js]: 👍

File: demo/DemoPopoverSection.jsx (hunk 5 above)
Comment (MACSkeptic) [inline demo/DemoPopoverSection.jsx]: should we build the demo as part of CI?

File: src/DataTable.jsx (hunk 1 above)
Comment (MACSkeptic) [inline src/DataTable.jsx]: or `ListTable` if we're following canon terminology

File: demo/DemoPopoverSection.jsx (hunk 5 above)
Comment (wpwood) [inline demo/DemoPopoverSection.jsx]: Maybe so. I found this when running `scripts/server` to run the demo.

File: src/DataTable.jsx (hunk 1 above)
Comment (wpwood) [inline src/DataTable.jsx]: I like using something other than just Table, since it could get confusing differentiating between a Table component and a table DOM element. I'm good with ListTable. I went with DataTable because it really does expect a collection of data (the whole cloning of children thing).

File: demo/DemoPopoverSection.jsx (hunk 5 above)
Comment (wpwood) [inline demo/DemoPopoverSection.jsx]: @adampickeral What do you think? I'm inclined _not_ to build the demo as part of the CI build, since that's building things we ship, and we don't ship the demo.

Comment (MACSkeptic) [inline demo/DemoPopoverSection.jsx]: I'm inclined to say we should build it as part of CI

these small mistakes seem very easy to make and go unnoticed, don't think any amount of code review would have caught it

File: src/DataTable.jsx (hunk 1 above)
Comment (MACSkeptic) [inline src/DataTable.jsx]: ^ I share your concern against just calling it Table, `Table` vs `table` is a very nuance distinction and might not be obvious amidst a more complicate component that uses this

File: demo/DemoPopoverSection.jsx (hunk 5 above)
Comment (adampickeral) [inline demo/DemoPopoverSection.jsx]: build it as part of CI

File: src/DataTable.jsx (hunk 3 above)
Comment (wpwood) [inline src/DataTable.jsx]: My concern with passing them as children is two-fold. First, I like the idea of just passing the children of ListTable to the `table`, and if the overlays were also children, we'd have to filter them out somehow. Second, we can't depend on the loading, error, and empty overlays to always be components of type `LoadingOverlay`, `ErrorOverlay`, and `EmptyOverlay`, since the user of the class may want to implement specialized behavior in their own overlay component and pass it in.

Comment (MACSkeptic) [inline src/DataTable.jsx]: the overlays could be just declarative wrappers (i.e.: just a div with the right classes with children) for the filtering

File: src/EmbeddedDataTable.jsx (hunk 9 above)
Comment (wpwood) [inline src/EmbeddedDataTable.jsx]: I may not, but we typically pass props through to the rendered component.

File: src/DataTable.jsx (hunk 3 above)
Comment (wpwood) [inline src/DataTable.jsx]: I guess I'm not clear what that would look like. The whole point of wrapping behavior in a component is to abstract it out so the caller doesn't need to know what's in it. It sounds like you're suggesting not doing that, in order to keep from passing in components as props, which we do quite a bit in canon-react. I don't always love it, but I think it makes sense here.

Can you show me an example of what this would look like?

Comment (MACSkeptic) [inline src/DataTable.jsx]: it would be a lot more code for this definitely... something like:

```javascript

let loadingOverlay = React.Children.toArray(this.props.children).find(c => c.type === LoadingOverlay);
let errorOverlay = React.Children.toArray(this.props.children).find(c => c.type === ErrorOverlay);
let emptyOverlay = React.Children.toArray(this.props.children).find(c => c.type === EmptyOverlay);
```

that would allow us to do something like this:

```javascript
<DataTable>
  <LoadingOverlay isLoading={true}>...</LoadingOverlay>
  <ErrorOverlay isErrored={true}>...</ErrorOverlay>
  <EmptyOverlay isEmpty={true}>...</EmptyOverlay>
  <DataTableBody>...</DataTableBody>
</DataTable>
```

I think it is totally debatable which one is better, I'm fine if we say we just go with props : )

File: demo/DemoListTableSection.jsx [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+    }
+  }
+
+  _compareNumericValues(a, b) {
+    if (a === b) {
+      return 0;
+    }
+
+    if (this.state.direction === SortDirection.ASCENDING) {
+      if (a < b) {
```
Comment (MACSkeptic) [inline demo/DemoListTableSection.jsx]: `return (a - b) / Math.abs(a - b);` ?

File: demo/DemoListTableSection.jsx [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+
+  _compareStringValues(a, b) {
+    if (this.state.direction === SortDirection.ASCENDING) {
+      return a.localeCompare(b);
+    } else {
+      return b.localeCompare(a);
+    }
+  }
+
+  _compareNumericValues(a, b) {
```
Comment (MACSkeptic) [inline demo/DemoListTableSection.jsx]: would this work for numeric values that are `'1', '2', '3'` instead of `1, 2, 3` (i.e.: numeric values, but as strings) ?

Comment (wpwood) [inline demo/DemoListTableSection.jsx]: Probably not. This is just the demo, so I didn't feel like it needed any extra bells and whistles.

Comment (MACSkeptic) [inline demo/DemoListTableSection.jsx]: didn't realise this was the demo 😊

**Comment by MACSkeptic** [comment]
:shipit:

**Comment by adampickeral** [comment]
would you mind adding some documentation about the public components (props needed, purpose, etc) being added and how they are used to the description of the PR?

**Comment by wpwood** [comment]
@adampickeral Do you have an example of what you'd like to see? It seems like the documentation we have is primarily the Demo classes, which I've provided. Also, it probably shouldn't reside in the PR, but the wiki is currently empty, so opinions on where it should go?

**Comment by adampickeral** [comment]
when we have a larger feature being added such as this, it's good to have documentation in the PR (or in an issue proposing the change) about the approach that is being taken. It makes reviewing the approach at a higher level easier, and it also makes it easier for anyone to update the official documentation after the feature is merged in.

https://github.com/rackerlabs/canon-react/pull/27
https://github.com/rackerlabs/canon-react/issues/55

**Comment by wpwood** [comment]
@adampickeral Gotcha. Thanks for the examples. I'll add something similar.

**Comment by adampickeral** [comment]
:+1:

File: package.json [hunk 14]
```diff
@@ -53,6 +53,7 @@
     "babelify": "7.2.0",
     "browserify": "<4.0.0",
     "browserify-shim": "3.8.12",
+    "enzyme": "2.3.0",
```
Comment (willcumbie) [inline package.json]: :+1:

File: src/ListTableBody.jsx [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        { this._cloneRowPerInstance() }
+      </tbody>
+    );
+  }
+
+  _cloneRowPerInstance() {
+    return this.props.collection.map(this._cloneRow);
+  }
+
+  _cloneRow(instance) {
```
Comment (willcumbie) [inline src/ListTableBody.jsx]: map() passes three arguments--the current item, the index, and the array as a whole. We should at least pass the first two into the key generator, it's not hard to imagine a scenario where the index might be needed there.

**Comment by willcumbie** [comment]
LGTM

File: demo/DemoListTableSection.jsx [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+    const sortedCollection = this._sort();
+
+    const tableStyle = {width: '500px'};
+
+    const emptyOverlay = (<EmptyOverlay title="Empty Overlay" subtitle="I am so empty" message="Empty, empty, empty" />);
+    const errorOverlay = (<ErrorOverlay message="There was an error loading data" />);
+    const loadingOverlay = (<LoadingOverlay />);
+
+    return (
+      <div className='rs-detail-section'>
```
Comment (hannahstwinter) [inline demo/DemoListTableSection.jsx]: Should we use Canon-React components in the demo? It would suck if something were updated/modified at some point and the demo broke.

File: src/ListTable.jsx [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+      loadingOverlay,
+      overlayStatus
+    } = this.props;
+
+    return (
+      <div { ...this.props } >
+        <table className="rs-list-table">
+          { children }
+        </table>
+        <ListTableOverlaySelector { ...{overlayStatus, emptyOverlay, errorOverlay, loadingOverlay} } />
```
Comment (hannahstwinter) [inline src/ListTable.jsx]: spacing nit - ```{ ...{ overlayStatus, emptyOverlay, errorOverlay, loadingOverlay } }```

File: src/TextColumnHeader.jsx [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+  label: '',
+  sortable: false,
+  onSort: () => {}
+};
+
+TextColumnHeader.propTypes = {
+  label: PropTypes.string.isRequired,
+  sortable: PropTypes.bool.isRequired,
+  onSort: PropTypes.func.isRequired,
+  direction: PropTypes.oneOf([
```
Comment (hannahstwinter) [inline src/TextColumnHeader.jsx]: PropTypes.oneOf(Object.keys(SortDirection))

File: demo/DemoListTableSection.jsx [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+    <CheckboxColumnHeader />
+    <TextColumnHeader columnId="name" className="name" sortable label="Name" />
+    <TextColumnHeader columnId="count" className="count" sortable label="Count" />
+  </ListTableHeader>
+);
+
+DemoListTableHeader.propTypes = {
+  onSort: React.PropTypes.func,
+  sortColumn: React.PropTypes.string,
+  direction: React.PropTypes.oneOf([
```
Comment (hannahstwinter) [inline demo/DemoListTableSection.jsx]: PropTypes.oneOf(Object.keys(SortDirection))

File: src/ListTableHeader.jsx [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+ListTableHeader.defaultProps = {
+  onSort: () => {}
+};
+
+ListTableHeader.propTypes = {
+  onSort: PropTypes.func,
+  sortColumn: PropTypes.string,
+  direction: PropTypes.oneOf([
```
Comment (hannahstwinter) [inline src/ListTableHeader.jsx]: PropTypes.oneOf(Object.keys(SortDirection))

File: src/ListTableRow.jsx [hunk 21]
```diff
@@ -0,0 +1,23 @@
+import { Component, PropTypes } from 'react';
+
+class ListTableRow extends Component {
```
Comment (hannahstwinter) [inline src/ListTableRow.jsx]: If this is stateful solely to handle the cloneChildren function as a class function I feel like it can be stored in a var just as easily. I know you </3 stateless components but I do think we should aim for consistency with when we choose to go which route. Refs seem like a reasonable delineator, and structuring logic and flow via class functions makes sense to me. This wouldn't convolute the latter, IMO, and doesn't touch the first point. Just a thought.

File: src/TextColumnHeader.jsx [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+class TextColumnHeader extends Component {
+  render() {
+    let classes;
+    const { label, sortable, direction, className } = this.props;
+
+    if (sortable) {
+      classes = classNames(
+        'rs-table-sort',
+        { 'rs-table-sort-asc': direction === SortDirection.ASCENDING },
+        { 'rs-table-sort-desc': direction === SortDirection.DESCENDING },
```
Comment (hannahstwinter) [inline src/TextColumnHeader.jsx]: you can make these one obj if you want -
```
{ 
  'rs-table-sort-asc': direction === SortDirection.ASCENDING,
  'rs-table-sort-desc': direction === SortDirection.DESCENDING 
}
```

File: src/TextColumnHeader.jsx [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+      </th>
+    );
+  }
+
+  _handleClick(ev) {
+    let newDirection;
+    const { direction } = this.props;
+
+    if (direction === SortDirection.ASCENDING) {
+      newDirection = SortDirection.DESCENDING;
```
Comment (hannahstwinter) [inline src/TextColumnHeader.jsx]: You could call this.props.onSort(SortDir.DESC) inline here and on 42 and save the var. It would read a bit better, IMO, less to track in my brain space - if ascending, sort the other way.

**Comment by hannahstwinter** [comment]
Great work, @wpwood ! This looks awesome. :+1:

File: demo/DemoListTableSection.jsx (hunk 16 above)
Comment (wpwood) [inline demo/DemoListTableSection.jsx]: Maybe so. It's like this in every demo that uses a section. I'll update it in a separate PR.
